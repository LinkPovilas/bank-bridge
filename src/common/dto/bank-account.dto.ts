import { Exclude, Expose } from 'class-transformer';

@Exclude()
export class BankAccountDto {
  @Expose()
  resourceId?: string;

  @Expose()
  name: string;

  @Expose()
  iban: string;

  @Expose()
  currency: string;

  constructor(data: BankAccountDto) {
    Object.assign(this, data);
  }
}
