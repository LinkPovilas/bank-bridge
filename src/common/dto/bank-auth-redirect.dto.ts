import { Exclude, Expose } from 'class-transformer';

@Exclude()
export class BankAuthRedirectDto {
  @Expose()
  readonly state: string;

  @Expose()
  readonly scaRedirect: string;

  constructor(data: BankAuthRedirectDto) {
    Object.assign(this, data);
  }
}
