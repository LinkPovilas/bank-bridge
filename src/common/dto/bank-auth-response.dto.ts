import { Exclude, Expose } from 'class-transformer';

@Exclude()
export class BankAuthDto {
  @Expose()
  readonly accessToken: string;

  @Expose()
  readonly refreshToken: string;

  @Expose()
  readonly scopes: string;

  @Expose()
  readonly accessTokenValidUntil: string;

  @Expose()
  readonly refreshTokenValidUntil: string;

  constructor(data: BankAuthDto) {
    Object.assign(this, data);
  }
}
