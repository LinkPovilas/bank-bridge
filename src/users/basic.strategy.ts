import {
  ForbiddenException,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { BasicStrategy as Strategy } from 'passport-http';
import { PassportStrategy } from '@nestjs/passport';
import { AuthService } from './auth.service';

@Injectable()
export class BasicStrategy extends PassportStrategy(Strategy) {
  constructor(private readonly authService: AuthService) {
    super();
  }

  async validate(username: string, password: string) {
    const user = await this.authService.validateUser(username, password);

    if (!user) {
      throw new UnauthorizedException();
    }

    if (!user.approved) {
      throw new ForbiddenException("User isn't approved.");
    }

    return user;
  }
}
