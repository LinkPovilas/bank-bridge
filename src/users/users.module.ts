import { Module } from '@nestjs/common';
import { UsersService } from './users.service';
import { UsersController } from './users.controller';
import { TypeOrmModule } from '@nestjs/typeorm';
import { User } from './entities/user.entity';
import { BasicStrategy } from './basic.strategy';
import { AuthService } from './auth.service';

@Module({
  imports: [TypeOrmModule.forFeature([User])],
  controllers: [UsersController],
  providers: [UsersService, AuthService, BasicStrategy],
})
export class UsersModule {}
