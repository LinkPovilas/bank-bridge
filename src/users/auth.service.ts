import { BadRequestException, Injectable } from '@nestjs/common';
import { UsersService } from './users.service';
import { CreateUserDto } from './dto/create-user.dto';
import { scrypt as _scrypt, randomBytes } from 'crypto';
import { promisify } from 'util';

const scrypt = promisify(_scrypt);

@Injectable()
export class AuthService {
  constructor(private readonly usersService: UsersService) {}

  async register(createUserDto: CreateUserDto) {
    const user = await this.usersService.findOne({
      where: { username: createUserDto.username },
    });

    if (user) {
      throw new BadRequestException('User already exists');
    }

    const salt = randomBytes(8).toString('hex');
    const hash = (await scrypt(createUserDto.password, salt, 32)) as Buffer;
    const result = `${salt}.${hash.toString('hex')}`;

    return this.usersService.create({ ...createUserDto, password: result });
  }

  async validateUser(username: string, password: string) {
    const user = await this.usersService.findOne({ where: { username } });

    if (!user) {
      return null;
    }

    const [salt, storedHash] = user.password.split('.');
    const hash = (await scrypt(password, salt, 32)) as Buffer;

    if (storedHash !== hash.toString('hex')) {
      return null;
    }

    const { password: _password, ...result } = user;
    return result;
  }
}
