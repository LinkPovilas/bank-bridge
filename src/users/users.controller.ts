import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
} from '@nestjs/common';
import { UsersService } from './users.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { ApiBasicAuth, ApiTags } from '@nestjs/swagger';
import { Public } from '../common/decorators/public.decorator';

@ApiTags('users')
@Controller('internal/users')
export class UsersController {
  constructor(private readonly usersService: UsersService) {}

  @Public()
  @Post()
  create(@Body() createUserDto: CreateUserDto) {
    return this.usersService.create(createUserDto);
  }

  @ApiBasicAuth()
  @Get()
  findAll() {
    return this.usersService.findAll();
  }

  @ApiBasicAuth()
  @Get(':id')
  findOne(@Param('id') id: number) {
    console.log(typeof id);
    return this.usersService.findOne({ where: { id } });
  }

  @ApiBasicAuth()
  @Patch(':id')
  update(@Param('id') id: number, @Body() updateUserDto: UpdateUserDto) {
    return this.usersService.update(id, updateUserDto);
  }

  @ApiBasicAuth()
  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.usersService.remove(+id);
  }
}
