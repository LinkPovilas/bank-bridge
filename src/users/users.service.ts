import { BadRequestException, Injectable } from '@nestjs/common';
import { CreateUserDto } from './dto/create-user.dto';
import { InjectRepository } from '@nestjs/typeorm';
import { FindOneOptions, Repository } from 'typeorm';
import { User } from './entities/user.entity';

@Injectable()
export class UsersService {
  constructor(
    @InjectRepository(User) private readonly userRepository: Repository<User>,
  ) {}

  create(createUserDto: CreateUserDto) {
    const user = this.userRepository.create(createUserDto);
    return this.userRepository.save(user);
  }

  findAll() {
    return this.userRepository.find();
  }

  findOne(options: FindOneOptions<User>) {
    return this.userRepository.findOne(options);
  }

  async update(id: number, attributes: Partial<User>) {
    const user = await this.findOne({ where: { id } });
    if (!user) {
      throw new BadRequestException('User not found');
    }
    return this.userRepository.save({ ...user, ...attributes });
  }

  async remove(id: number) {
    const user = await this.findOne({ where: { id } });
    if (!user) {
      throw new BadRequestException('User not found');
    }
    return this.userRepository.remove(user);
  }
}
