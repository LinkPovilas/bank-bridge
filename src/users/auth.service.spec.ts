import { Test, TestingModule } from '@nestjs/testing';
import { AuthService } from './auth.service';
import { UsersService } from './users.service';
import { Role } from '../common/enums/role.enum';
import { CreateUserDto } from './dto/create-user.dto';
import { FindOneOptions } from 'typeorm';
import { User } from './entities/user.entity';

describe('AuthService', () => {
  let service: AuthService;
  let mockUsersService: Partial<UsersService>;

  const testUser: CreateUserDto = {
    email: 'test@test.eu',
    username: 'test',
    password: 't35t',
  };

  beforeEach(async () => {
    mockUsersService = {
      findOne: jest.fn((options: FindOneOptions<User>) =>
        Promise.resolve({
          id: Math.floor(Math.random() * 999999),
          email: 'test@test.eu',
          username: options.where['username'],
          password: 't35t',
          role: Role.SERVICE,
          approved: false,
        }),
      ),
      create: jest.fn((createUserDto: CreateUserDto) =>
        Promise.resolve({
          id: Math.floor(Math.random() * 999999),
          ...createUserDto,
          role: Role.SERVICE,
          approved: false,
        }),
      ),
    };

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        AuthService,
        { provide: UsersService, useValue: mockUsersService },
      ],
    }).compile();

    service = module.get<AuthService>(AuthService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('register', () => {
    it('should create a new user', async () => {
      jest
        .spyOn(mockUsersService, 'findOne')
        .mockReturnValueOnce(Promise.resolve(null));

      const user = await service.register(testUser);

      expect(mockUsersService.findOne).toHaveBeenCalled();
      expect(mockUsersService.create).toHaveBeenCalled();

      expect(user).toBeDefined();
      expect(user).toMatchSnapshot({
        id: expect.any(Number),
        password: expect.any(String),
      });

      expect(user.password).not.toBe(testUser.password);
      const [salt, hash] = user.password.split('.');
      expect(salt).toBeDefined();
      expect(hash).toBeDefined();
    });

    it('should throw an error if user already exists', async () => {
      await expect(service.register(testUser)).rejects.toThrow();
    });
  });

  describe('validateUser', () => {
    let registeredUser: User;

    beforeEach(async () => {
      jest
        .spyOn(mockUsersService, 'findOne')
        .mockReturnValueOnce(Promise.resolve(null));

      registeredUser = await service.register(testUser);
    });

    it('should return user if credentials are valid', async () => {
      jest
        .spyOn(mockUsersService, 'findOne')
        .mockReturnValueOnce(Promise.resolve(registeredUser));

      const user = await service.validateUser(
        testUser.username,
        testUser.password,
      );

      expect(mockUsersService.create).toHaveBeenCalled();
      expect(mockUsersService.findOne).toHaveBeenCalledTimes(2);

      expect(user).toBeDefined();
      expect(user).toMatchSnapshot({
        id: expect.any(Number),
      });
    });

    it('should return null if password is invalid', async () => {
      jest
        .spyOn(mockUsersService, 'findOne')
        .mockReturnValueOnce(Promise.resolve(registeredUser));

      const user = await service.validateUser(testUser.username, 'password');

      expect(mockUsersService.findOne).toHaveBeenCalledTimes(2);
      expect(user).toBeNull();
    });

    it('should return null if username is invalid', async () => {
      jest
        .spyOn(mockUsersService, 'findOne')
        .mockReturnValueOnce(Promise.resolve(null));

      const user = await service.validateUser('username', testUser.password);

      expect(mockUsersService.findOne).toHaveBeenCalledTimes(2);
      expect(user).toBeNull();
    });
  });
});
