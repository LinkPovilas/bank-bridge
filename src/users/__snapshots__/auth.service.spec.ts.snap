// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService register should create a new user 1`] = `
{
  "approved": false,
  "email": "test@test.eu",
  "id": Any<Number>,
  "password": Any<String>,
  "role": "service",
  "username": "test",
}
`;

exports[`AuthService validateUser should return user if credentials are valid 1`] = `
{
  "approved": false,
  "email": "test@test.eu",
  "id": Any<Number>,
  "role": "service",
  "username": "test",
}
`;
